### ISSUE TITLE
🐞 Bug

### ISSUE THREAD

**Issue by G-K-404** [CircuitVerse/CircuitVerse#5212]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

A bug in all the flipflops is that when both async reset and preset are set to high, the output is set to high, but in theory, the flip flop goes into a race condition similar to when both S and R are set to high, since the state(Q) remains at previous state and does not change to a new one, the state behavior must be the same for when both async preset and reset are high.

### Steps To Reproduce

1. Open the simulator in Circuit Verse.
2. select a flip flop of choice.
3. select the input bit states to be your desired values.
4. set the async preset and reset to 1.

### Expected Behavior

As mentioned in the description, it would be better to maintain a standard state behavior for race. This can be achieved by either setting the output state to the previous state or displaying an error saying "Race condition" when race invoking inputs are given, these changes will help us to ensure that no misinformation is given.

### Screenshots

![image](https://github.com/user-attachments/assets/d6087630-521d-4631-a6fd-9feb3ac086e3)
above is the state behavior when S and R are 1.
![image](https://github.com/user-attachments/assets/2e7f2230-9e64-422e-a0ca-51018cfe4db9)
above is the state behavior when S and R are 0 and 1 and both async reset and preset are 1.

### Device Information [optional]

```markdown
- OS: Windows11
- Browser: Opera
```

### Additional context

creating an error message saying "Race condition" will require changing structure of many components, but a try will be made at trying to make it work

### Are you working on this issue?

Yes

**Comment by salmoneatenbybear** [comment]
@gr455 @092vk can you please check if this exists ?

**Comment by salmoneatenbybear** [comment]
@G-K-404 can you please name the bug issue appropriately

**Comment by 092vk** [comment]
@salmoneatenbybear This is a valid bug, we should raise contentions on race conditions. the solution will involve raising contentions whenever race conditions arise. Also the behaviour of race conditions are not uniform , for example in case of SR being 1,1 the race condition is handled differently then when present and reset is 1,1.

**Comment by salmoneatenbybear** [comment]
@092vk can you fix this in a pr ?

**Comment by 092vk** [comment]
> [@092vk](https://github.com/092vk) can you fix this in a pr ?

working on it

**Comment by G-K-404** [comment]
> @salmoneatenbybear This is a valid bug, we should raise contentions on race conditions. the solution will involve raising contentions whenever race conditions arise. Also the behaviour of race conditions are not uniform , for example in case of SR being 1,1 the race condition is handled differently then when present and reset is 1,1.

@092vk Yea that's right, there's some more race conditions like JK being 11 for a long time, you will have to keep toggling the output. The behaviour of circuits during different races is surely different but, the behaviour of the circuit during the async preset and reset race can be solved by letting the output remain the same or by restricting the user from giving both async reset and preset at once. And one more point to be noted is that in an actual SR flip flop the clear and preset are taken as active low, considering all these will help in making a better and more accurate working of the flip flops. Kudos!

**Comment by 092vk** [comment]
@G-K-404 Yeah , actually the reset and preset in majority of flip flops are active low , but active high is also used . But can you confirm if preset is working fine , I checked it and found it not working . Reset is working fine.

**Comment by salmoneatenbybear** [comment]
@G-K-404 would you like to be assigned and collaborate with @092vk ?

**Comment by G-K-404** [comment]
> @G-K-404 Yeah , actually the reset and preset in majority of flip flops are active low , but active high is also used . But can you confirm if preset is working fine , I checked it and found it not working . Reset is working fine.

@092vk In the actual code? I think I remember being satisfied by the preset I'm not quite sure though.

**Comment by G-K-404** [comment]
> @G-K-404 would you like to be assigned and collaborate with @092vk ?
@salmoneatenbybear Yeah, sure I'd be really happy about that!

**Comment by G-K-404** [comment]
> @G-K-404 would you like to be assigned and collaborate with @092vk ?
@salmoneatenbybear Yeah, sure I'd be really happy about that!

**Comment by 092vk** [comment]
Na , when reset = 0 and preset =1 , no mater other conditions the Q=1, right ? , but this is not happening

**Comment by 092vk** [comment]
> > [@G-K-404](https://github.com/G-K-404) would you like to be assigned and collaborate with [@092vk](https://github.com/092vk) ?
> > [@salmoneatenbybear](https://github.com/salmoneatenbybear) Yeah, sure I'd be really happy about that!

Message me on slack , lets discuss it and solve it together

**Comment by 092vk** [comment]
@G-K-404 My Discord id : vk092

**Comment by G-K-404** [comment]
> @G-K-404 My Discord id : vk092

@092vk Mine is gk_69.69 a friend request has been sent, please do accept it.

**Comment by ThatDeparted2061** [comment]
From the overview of this convo.
Theoritically reset and preset shudnt be high together but in reality it might happen and to prevent that we have preset with higher precendence over reset. This is okay and need not be changed
Ps: correct me if i read the issue wrong

**Comment by G-K-404** [comment]
> From the overview of this convo.
> Theoritically reset and preset shudnt be high together but in reality it might happen and to prevent that we have preset with higher precendence over reset. This is okay and need not be changed
> Ps: correct me if i read the issue wrong

@ThatDeparted2061 You might have read the code wrong. The code gives higher precedence to preset only when reset is toggled that means you don't have a preset in the actual circuit it is toggled only when reset is high which is pretty much an example of a wrong simulation.

**Comment by ThatDeparted2061** [comment]
Race conditions dont happen with reset and preset working at the same time, it happens when the change in input causes two different end states.
Reset and preset both high is an error that shudnt happen in your circuit and is not a race condition but rather a circuit short coming

**Comment by ThatDeparted2061** [comment]
This is something I am sure of
> Race conditions dont happen with reset and preset working at the same time, it happens when the change in input causes two different end states.
> Reset and preset both high is an error that shudnt happen in your circuit and is not a race condition but rather a circuit short coming

**Comment by ThatDeparted2061** [comment]
So instead of calling reset and preset high at the same time as race condition (it is theoritically wrong) it shud be circuit short coming
Race conditions dont happen with reset and preset working at the same time, it happens when the change in input causes two different end states.
Reset and preset both high is an error that shudnt happen in your circuit and is not a race condition but rather a circuit short coming

**Comment by ThatDeparted2061** [comment]
@092vk 
Your PR s are wrong, 
### Reset and preset high at the same time is an invalid state not a race 

A race condition occurs when Two or more signals change simultaneously in a way that the final state depends on the exact timing of these changes

**Comment by 092vk** [comment]
> @092vk 
> Your PR s are wrong, 
> ### Reset and preset high at the same time is an invalid state not a race 
> 
> A race condition occurs when Two or more signals change simultaneously in a way that the final state depends on the exact timing of these changes

Let's have a meet and discuss it

**Comment by ThatDeparted2061** [comment]
Yeah better

**Comment by Me-Priyank** [comment]
Outdsted. Simulator has been moved to cv-frontend-vue repo
